charts/spire/charts/spire-agent/templates/daemonset.yaml

```diff
@@ -43,6 +43,22 @@ spec:
           args: ["-t", "30", "-h", "{{ include "spire-agent.server-address" . | trim }}", "-p", {{ .Values.server.port | quote }}]
           resources:
             {{- toYaml .Values.waitForIt.resources | nindent 12 }}
+        {{- if gt (int (dig "fsGroup" 0 .Values.podSecurityContext)) 0 }}
+        - name: fsgroupfix
+          image: {{ template "spire-lib.image" (dict "image" .Values.fsGroupFix.image "global" .Values.global) }}
+          imagePullPolicy: {{ .Values.fsGroupFix.image.pullPolicy }}
+          command: ["bash", "-c"]
+          args:
+            - "chown -R {{ .Values.podSecurityContext.runAsUser }}:{{ .Values.podSecurityContext.fsGroup }} {{ include "spire-agent.socket-path" . | dir }}"
+          resources:
+            {{- toYaml .Values.fsGroupFix.resources | nindent 12 }}
+          volumeMounts:
+            - name: spire-agent-socket-dir
+              mountPath: {{ include "spire-agent.socket-path" . | dir }}
+          securityContext:
+            runAsUser: 0
+            runAsGroup: 0
+        {{- end }}
```

Comment:
Please check if we can do the following:

Remove the configurable for the spire-agent mountPath to use the image default.

https://github.com/spiffe/spire/blob/main/Dockerfile#L52-L54

This image default already has the right permissions. In that case we don't even need the configurable for running as root. Then we should just always run as non-root. Don't see a reason to run as root if we can by default run as non-root.

Probably remove this configurable so people can't break the permissions that should be good by default.

```
# -- The unix socket path to the spire-agent
socketPath: /run/spire/agent-sockets/spire-agent.sock
```

Reply:
Seems like we have a few 2 options here regarding the configurable:
1. Leave the configurable in place but default it to `/tmp/spire-agent/public/api.sock` with note that changing it will require running as root
2. Remove configurable altogether. Is it needed?

Regarding running as root, I think we can still support that with the `runAsUser` configurable but having the default be `1000`?

Reply:
I would be happy if the configurable is not needed. I have not been able to prove that is it not needed. Someone needs to step up and find an alternate solution if this one is unacceptable.

Reply:
Wait... I'm getting confused. there are lots of configurables at this point being discussed. Lets recap so we're all on the same page.

The agent socket needs to be configurable, as it needs to be exposed on a hostpath. It isn't possible to unhardcode that part and support nested spire or kubernetes clusters that don't install kubelet in the default location.

It may be possible to hardcode its mountpoint in the container. but that doesn't affect the permissions of that mount. The permissions come from the host/kubelet. Not sure what the benefit to that change is.

This pr changes the permissions to match what is specified by the configuration setting that a user can optionally set in the securityContext, if they have done so. This fixes the bug.

And, this pr contains the recommendation in the example pr to run as nonroot. But the user can override it back to root if needed. The default for the chart, when not using the production recommendations is to run as root.

I'm not seeing any alternative to this pr that will work. If someone wants to submit a different pr, with a workable solution for all of these things, I'd love to see one.

Reply:
Got it, does this make sense then:

- Change default of socket path to tmp dir
- Change default of runAsUser to 1000 or some non root value
- If socket patch is changed and runAsUser is non root, run the init container

Reply:
Will check in coming week to see if I can get it to work. This week was a busy week.